index.bs

```diff
@@ -261,39 +259,78 @@ vrDisplay.requestPresent([{ source: canvas }]).then(function() {
 
 ## VRLayer ## {#interface-vrlayer}
 
-The {{VRLayer}} interface is provided to a {{VRDisplay}} and presented in the HMD.
+The {{VRLayer}} is the base interface for the various layer types that can be
+provided to a {{requestPresent()}} call to be presented in the HMD.
 
-<pre class="idl" id="vrlayer-dictionary">
+<pre class="idl" id="vrlayer-interface">
+interface VRLayer {};
+</pre>
+
+## CanvasVRLayer ## {#interface-canvasvrlayer}
+
+The {{CanvasVRLayer}} describes a layer containing head tracked stereo content
+rendered to a canvas with a WebGL context.
+
+<pre class="idl" id="canvasvrlayer-interface">
 typedef (HTMLCanvasElement or
-         OffscreenCanvas) VRSource;
+         OffscreenCanvas) CanvasVRSource;
 
-[Constructor(optional VRLayerInit layer)]
-interface VRLayer {
-  readonly attribute VRSource? source;
+[Constructor(optional CanvasVRSource source)]
+interface CanvasVRLayer : VRLayer {
+  attribute CanvasVRSource source;
 
-  readonly attribute sequence&lt;float&gt; leftBounds;
-  readonly attribute sequence&lt;float&gt; rightBounds;
-};
+  attribute float renderScale;
+  attribute boolean resizeSourceOnPresent;
+
+  readonly attribute unsigned long renderWidth;
+  readonly attribute unsigned long renderHeight;
 
-dictionary VRLayerInit {
-  VRSource? source = null;
+  void setLeftBounds(float left, float bottom, float right, float top);
+  sequence&lt;float&gt; getLeftBounds();
 
-  sequence&lt;float&gt; leftBounds = [ ];
-  sequence&lt;float&gt; rightBounds = [ ];
+  void setRightBounds(float left, float bottom, float right, float top);
+  sequence&lt;float&gt; getRightBounds();
 };
 </pre>
 
-### Attributes ### {#vrlayer-attributes}
+### Attributes ### {#canvasvrlayer-attributes}
+
+<dfn attribute for="CanvasVRLayer">source</dfn>
+The {{CanvasVRLayer/source}} attribute defines the canvas whose contents will be presented for this layer by the {{VRDisplay}} when {{VRDisplay}}.{{submitFrame()}} is called. Upon being passed into {{requestPresent()}} the current backbuffer of the source's context MAY be lost, even if the context was created with the <code>preserveDrawingBuffer</code> context creation attribute set to <code>true</code>. Calls to {{requestPresent()}} MUST be rejected if the {{CanvasVRLayer/source}} of any {{CanvasVRLayer}} provided does not represent a canvas with a <code>webgl</code> or <code>webgl2</code> context.
+
```

Comment:
This actually wasn't aimed at a multiadapter scenario, but instead accounts for cases where the render target for the canvas needs to be swapped out for something VR specific. If we were attempting to strictly adhere to preserveDrawingBuffer semantics that would require a copy of the existing content over to the new render target, something that will be lost 99% of the time anyway because the canvas will likely be resized to match the VRDisplay size. So this text simply allows the UA to not guarantee that the render target content will be preserved.

The multi-adapter scenario does need to be accounted for, though. And unfortunately I'm still not sure where or how. :(